src/daemon/daemon.cpp

```diff
@@ -2668,6 +2680,82 @@ catch (const std::exception& e)
     status_promise->set_value(grpc::Status(grpc::StatusCode::INTERNAL, e.what(), ""));
 }
 
+void mp::Daemon::clone(const CloneRequest* request,
+                       grpc::ServerReaderWriterInterface<CloneReply, CloneRequest>* server,
+                       std::promise<grpc::Status>* status_promise)
+{
+    try
+    {
+        config->factory->require_clone_support();
+        mpl::ClientLogger<CloneReply, CloneRequest> logger{mpl::level_from(request->verbosity_level()),
+                                                           *config->logger,
+                                                           server};
+
+        const auto& source_name = request->source_name();
+        const auto [instance_trail, status] = find_instance_and_react(operative_instances,
+                                                                      deleted_instances,
+                                                                      source_name,
+                                                                      require_operative_instances_reaction);
+        if (status.ok())
+        {
+            const std::string destination_name = generate_destination_instance_name_for_clone(*request);
+
+            auto rollback_clean_up_all_resource_of_dest_instance = sg::make_scope_guard(
+                [this, destination_name]() noexcept -> void { release_resources(destination_name); });
+
+            assert(instance_trail.index() == 0);
+            const auto source_vm_ptr = std::get<0>(instance_trail)->second;
+            assert(source_vm_ptr);
+            const VirtualMachine::State source_vm_state = source_vm_ptr->current_state();
+            if (source_vm_state != VirtualMachine::State::stopped && source_vm_state != VirtualMachine::State::off)
+            {
+                throw std::runtime_error("Please stop instance " + source_name + " before you clone it.");
```

Comment:
Instead of throwing an exception, you could just preemptively return with the `FAILED_PRECONDITION` status.

Reply:
Can be done.